parser/impl/parser.cpp

```diff
@@ -469,17 +469,201 @@ sPtr<ast::IfStatement> Parser::parseIfStatement() {
 // ---- @aabounegm
 
 sPtr<ast::Expression> Parser::parseExpression() {
-    // TODO: Concrete implementation
+    return parseBinaryExpression(0);
+}
+
+int Parser::opPrec(TokenType token) {
+    switch (token) {
+    case TokenType::Or:  // or
+    case TokenType::Xor: // xor
+        return 1;
+    case TokenType::And: // and
+        return 2;
+    case TokenType::Eq:      // =
+    case TokenType::Neq:     // /=
+    case TokenType::Less:    // <
+    case TokenType::Leq:     // <=
+    case TokenType::Greater: // >
+    case TokenType::Geq:     // >=
+        return 3;
+    case TokenType::Add: // +
+    case TokenType::Sub: // -
+        return 4;
+    case TokenType::Mul: // *
+    case TokenType::Div: // /
+    case TokenType::Mod: // %
+        return 5;
+    case TokenType::Not:       // not
+    case TokenType::OpenParen: // ()
+        return 6;
+    case TokenType::Dot:       // .
+    case TokenType::OpenBrack: // []
+        return 7;
+    default:
+        return -1; // unknown op
+    }
+}
+
+bool Parser::isPrimitive(TokenType token) {
+    return token == TokenType::Identifier || token == TokenType::Int ||
+           token == TokenType::Real || token == TokenType::True ||
+           token == TokenType::False;
+}
+
+sPtr<ast::Expression> Parser::parseUnaryExpression() {
+    Token tok = m_lexer.Peek();
+    ast::Expression expr;
+    if (opPrec(tok.type) >= 0) {
+        if (tok.type == TokenType::Not || tok.type == TokenType::Sub ||
+            tok.type == TokenType::Add) {
+
+            tok = m_lexer.Next();
+            expr.operation = tok;
+            expr.operand1 = parseUnaryExpression();
+            return std::make_shared<ast::Expression>(expr);
+
+        } else if (tok.type == TokenType::OpenParen) {
+            tok = m_lexer.Next();
+            sPtr<ast::Expression> expr = parseBinaryExpression(0);
+            if (m_lexer.Peek().type == TokenType::CloseParen) {
+                tok = m_lexer.Next();
+                return expr;
+            }
+            // error
+            m_errors.push_back(Error{
+                .pos = tok.pos,
+                .message = "Expected to find ')'",
+            });
+            return nullptr;
```

Comment:
The order feels a little atypical 😅. Usually, we check for an error and break the flow (with a `return`) if so, and continue the normal logic otherwise.
```suggestion
            if (m_lexer.Peek().type != TokenType::CloseParen) {
                m_errors.push_back(Error{
                    .pos = tok.pos,
                    .message = "Expected to find ')'",
                });
                while (m_lexer.Peek().type != TokenType::CloseParen) m_lexer.Next();
                return nullptr;
            }
            tok = m_lexer.Next();
            return expr;
```
Also, don't store the value of `m_lexer.Next()` if you just want to consume it but not actually use it because it becomes confusing to read.